drivers/counter/counter_smartbond_timer.c

```diff
@@ -0,0 +1,375 @@
+	bool absolute = alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE;
+	uint32_t flags = alarm_cfg->flags;
+	uint32_t val = alarm_cfg->ticks;
+	bool irq_on_late;
+	int err = 0;
+	uint32_t max_rel_val;
+	uint32_t now;
+	uint32_t diff;
+
+	if (chan != 0 || alarm_cfg->ticks > counter_smartbond_get_top_value(dev)) {
+		return -EINVAL;
+	}
+
+	if (data->callback) {
+		return -EBUSY;
+	}
+
+	now = counter_smartbond_read(dev);
+	data->callback = alarm_cfg->callback;
+	data->user_data = alarm_cfg->user_data;
+
+	__ASSERT_NO_MSG(data->guard_period < TIMER_TOP_VALUE);
+
+	if (absolute) {
+		max_rel_val = TIMER_TOP_VALUE - data->guard_period;
+		irq_on_late = flags & COUNTER_ALARM_CFG_EXPIRE_WHEN_LATE;
+	} else {
+		/* If relative value is smaller than half of the counter range
+		 * it is assumed that there is a risk of setting value too late
+		 * and late detection algorithm must be applied. When late
+		 * setting is detected, interrupt shall be triggered for
+		 * immediate expiration of the timer. Detection is performed
+		 * by limiting relative distance between CC and counter.
+		 *
+		 * Note that half of counter range is an arbitrary value.
+		 */
+		irq_on_late = val < (TIMER_TOP_VALUE / 2U);
+		/* limit max to detect short relative being set too late. */
+		max_rel_val = irq_on_late ? TIMER_TOP_VALUE / 2U : TIMER_TOP_VALUE;
+		val = (now + val) & TIMER_TOP_VALUE;
+	}
+	timer->TIMER2_RELOAD_REG = val;
+	*timer_clear_irq_reg = 1;
+	/* decrement value to detect also case when val == counter_smartbond_read(dev). Otherwise,
+	 * condition would need to include comparing diff against 0.
+	 */
+	diff = ((val - 1U) - counter_smartbond_read(dev)) & TIMER_TOP_VALUE;
+	if (diff > max_rel_val) {
+		if (absolute) {
+			err = -ETIME;
+		}
+
+		/* Interrupt is triggered always for relative alarm and
+		 * for absolute depending on the flag.
+		 */
+		if (irq_on_late) {
+			NVIC_SetPendingIRQ(config->irqn);
+		} else {
+			data->callback = NULL;
+		}
+	} else {
+		if (diff == 0) {
+			/* RELOAD value could be set just in time for interrupt
+			 * trigger or too late. In any case time is interrupt
+			 * should be triggered. No need to enable interrupt
+			 * on TIMER just make sure interrupt is pending.
+			 */
+			NVIC_SetPendingIRQ(config->irqn);
+		} else {
+			timer->TIMER2_CTRL_REG |= TIMER2_TIMER2_CTRL_REG_TIM_IRQ_EN_Msk;
+		}
+	}
+
+	return err;
+}
+
+static int counter_smartbond_cancel_alarm(const struct device *dev, uint8_t chan)
+{
+	const struct counter_smartbond_config *config = dev->config;
+	TIMER2_Type *timer = config->timer;
+	struct counter_smartbond_data *data = dev->data;
+
+	ARG_UNUSED(chan);
+
+	timer->TIMER2_CTRL_REG &= ~TIMER2_TIMER2_CTRL_REG_TIM_IRQ_EN_Msk;
+	data->callback = NULL;
+
+	return 0;
+}
+
+static int counter_smartbond_set_top_value(const struct device *dev,
+					   const struct counter_top_cfg *cfg)
+{
+	ARG_UNUSED(dev);
+
+	if (cfg->ticks != 0xFFFFFF) {
+		return -ENOTSUP;
+	}
+
+	return 0;
```

Comment:
I think it would make more sense to always return `-ENOTSUP` since top value is fixed